src/thermo/Phase.cpp

```diff
@@ -628,21 +715,43 @@ doublereal Phase::elementalMoleFraction(const size_t m) const
     return numerator / denom;
 }
 
-doublereal Phase::molarDensity() const
+double Phase::molarDensity() const
 {
     return density()/meanMolecularWeight();
 }
 
-void Phase::setMolarDensity(const doublereal molar_density)
+void Phase::setMolarDensity(const double molar_density)
 {
+    assertCompressible("setMolarDensity");
     m_dens = molar_density*meanMolecularWeight();
 }
 
-doublereal Phase::molarVolume() const
+double Phase::molarVolume() const
 {
     return 1.0/molarDensity();
 }
 
+void Phase::setDensity(const double density_)
+{
+    assertCompressible("setDensity");
+    if (density_ > 0.0) {
+        m_dens = density_;
+    } else {
+        throw CanteraError("Phase::setDensity()",
+            "density must be positive. density = {}", density_);
+    }
+}
+
+void Phase::assignDensity(const double density_)
+{
+    if (density_ >= 0.0) {
+        m_dens = density_;
```

Comment:
Why is density = 0 allowed for `assignDensity`, when `setDensity` requires a strictly positive value?

Reply:
Yup. One of the unit tests uses a Phase with zero density. Since I wasn't sure whether this was by design, I changed things here.

Reply:
Where is this test? How was it previously able to set the density to zero?

Reply:
@speth It fails in `Kinetics.ElectrochemFromYaml`, where `LatticePhase` is instantiated with `m_site_density` equal to 0.
```C++
assignDensity(meanMolecularWeight() * m_site_density);
```
raises the error. What's the most appropriate way: changing the default `m_site_density` or using a 'small' number to avoid the error?

Reply:
Oh, that's interesting. I would say that zero site density in LatticePhase is as much of a problem as zero density in other phases, so we should probably initialize it to something like `SmallNumber` in the first place. The `surface-phases.yaml` file should also probably specify `site-density` for those phase entries.

Looking at this further, this stems from a loophole in `Phase::setMolarDensity` which sets `m_dens` directly but doesn't impose any requirements. It's always something...

Reply:
Yes, it’s always something with this PR 😁 . Suggesting the fix indicated above (setting density to a small number if site-density is zero), in absence of clear physical intuition here. Any solution should allow YAML not specifying a site density ...

Reply:
I'm happy to defer any further digging on this to later. Your fix for the immediate issue seems fine.